warehouse/oidc/forms/activestate.py

```diff
@@ -0,0 +1,213 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+from collections.abc import Callable
+from typing import Any, TypedDict
+
+import requests
+import sentry_sdk
+import wtforms
+
+from warehouse import forms
+from warehouse.i18n import localize as _
+from warehouse.oidc.forms._core import PendingPublisherMixin
+
+_VALID_ORG_URL_NAME_AND_ACTOR_NAME = re.compile(r"^[a-zA-Z0-9-]{3,40}$")
+_VALID_PROJECT_NAME = re.compile(r"^[.a-zA-Z0-9-]{3,40}$")
+_DOUBLE_DASHES = re.compile(r"--+")
+
+_ACTIVESTATE_GRAPHQL_API_URL = "https://platform.activestate.com/graphql/v1/graphql"
+_GRAPHQL_GET_ORGANIZATION = "query($orgname: String) {organizations(where: {display_name: {_eq: $orgname}}) {added}}"  # noqa
+_GRAPHQL_GET_ACTOR = (
+    "query($username: String) {users(where: {username: {_eq: $username}}) {user_id}}"
+)
+_TIMEOUT = 5
+
+
+class UserResponse(TypedDict):
+    user_id: str
+
+
+class GqlResponse(TypedDict):
+    data: dict[str, Any]
+    errors: list[dict[str, Any]]
+
+
+def _no_double_dashes(form, field):
+    if _DOUBLE_DASHES.search(field.data):
+        raise wtforms.validators.ValidationError(
+            _("Double dashes are not allowed in the name")
+        )
+
+
+def _no_leading_or_trailing_dashes(form, field):
+    if field.data.startswith("-") or field.data.endswith("-"):
+        raise wtforms.validators.ValidationError(
+            _("Leading or trailing dashes are not allowed in the name")
+        )
```

Comment:
Thought, nonblocking: I think this could be worked into the regexps themselves, rather than having a separate validator. But IMO this is not worth trying to do now, something we can iterate on in the future 🙂

Reply:
I think I'm going to skip this as I actually prefer if we had more isolated validators that can give specific feedback.  The current regex, if it were to give helpful feedback, needs to state multiple things as the possible fix.  With this specific validator, it can state exactly what's wrong.

Reply:
I don't think I totally agree here.  I'm providing the granular regex's checks to provide a better user experience in this particular context but we could get away with "That organization does not exist".  Then the user would double check what they entered.

Maybe that's the better solution.  The user isn't trying to come up with an org name, they know the org.  They just need feedback that they entered the wrong thing, ie. "Not found"

So I would remove the regex checks on the format of the org name, then it falls to `_lookup_organization` then the empty list response is caught there.

What do you think @miketheman or @woodruffw ?